rfcs/text/0005_route_handler.md

```diff
@@ -0,0 +1,184 @@
+
+#### KibanaResponseToolkit methods
+Basic primitives:
+```typescript
+type HttpResponsePayload = undefined | string | JSONValue | Buffer | Stream;
+interface HttpResponseOptions {
+  headers?: {
+   // list of  known headers
+   ...
+   // for custom headers:
+   [header: string]: number | string | string[] | undefined;
+  }
+}
+
+type StatusCode = 200 | 202 | 204 | 302 | 400 | 401 | 403 | 404 | 409 | 500;
+```
+
+##### Success
+Server indicated that request was accepted:
+```typescript
+type SuccessResponse<T> = <T extends HttpResponsePayload>(
+  payload: T,
+  options?: HttpResponseOptions
+) => KibanaResponse<T>;
+
+const kibanaResponseToolkit = {
+  ok: <T extends HttpResponsePayload>(payload: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(200, payload, options),
+  accepted: <T extends HttpResponsePayload>(payload: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(202, payload, options),
+  noContent: (options?: HttpResponseOptions) => new KibanaResponse(204, undefined, options)
+```
+
+##### Redirection
+The server wants a user to perform additional actions.
+```typescript
+const kibanaResponseToolkit = {
+  redirected: (url: string, options?: HttpResponseOptions) => new KibanaResponse(302, url, options),
+  notModified: (options?: HttpResponseOptions) => new KibanaResponse(304, undefined, options),
+```
+
+#####  Error
+Server signals that request cannot be handled and explains details of the error situation
+```typescript
+// Supports attaching additional data to send to the client
+interface ResponseError extends Error {
+  meta: {
+    data?: JSONValue;
+    code?: string; // error code to simplify search, translations in i18n, etc.
+    link?: string; // link to the docs
+  }
+}
+
+const kibanaResponseToolkit = {
+  // Client errors
+  badRequest: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(400, err, options),
+  unauthorized: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(401, err, options),
+
+  forbidden: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(403, err, options),
+  notFound: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(404, err, options),
+  conflict: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(409, err, options),
+
+  // Server errors
+  internal: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(500, err, options),
+```
+
+##### Generic
+If a custom respond is required
+```typescript
+interface GenericOptions extends HttpResponseOptions {
+  statusCode: StatusCode;
+}
+export const kibanaResponseToolkit = {
+  generic: <T extends HttpResponsePayload>(payload: T, {statusCode, ...options}: GenericOptions) =>
+    new KibanaResponse(statusCode, payload, options),
```

Comment:
Not a huge fan of this name, it makes this sound special in some way beyond just customization. I'd prefer something more like `t.response`.

Reply:
The main objective of the helpers is a validation of a result and type checking. From this point of view, a `generic` or `custom` helper is a special case, indeed. I'm open to renaming, though. How about `custom`?

Reply:
`custom` sounds better 👍